controller/src/controller.rs

```diff
@@ -319,15 +348,120 @@ impl<T: BottlerocketShadowClient> BrupopController<T> {
+                    continue;
+                }
+            }
+
+            // Cleanup BRS when the operator is removed from a node
+            let brss_name = self
+                .all_brss()
+                .into_iter()
+                .map(|brs| brs.name_any())
+                .collect();
+            let nodes = self.all_nodes();
+            self.bottlerocketshadows_cleanup(nodes, brss_name).await?;
+
+            // Emit metrics at the end of the loop in case the loop didn't progress any nodes.
+            self.emit_metrics()?;
+
+            // Sleep until it's time to check for more action.
+            sleep(ACTION_INTERVAL).await;
+        }
+    }
+
+    /// Runs the event loop for the Brupop controller.
+    ///
+    /// Because the controller wants to gate the number of simultaneously updating nodes, we can't allow the update state machine
+    /// of each individual bottlerocket node to run concurrently and in an event-driven fashion, as is typically done with controllers.
+    /// Instead, we will keep an updated store of `BottlerocketShadow` objects based on cluster events, and then periodically make
+    /// scheduling decisions based on that store.
+    ///
+    /// The controller is designed to run on a single node in the cluster and rely on the scheduler to ensure there is always one
+    /// running; however, it could be expanded using leader-election and multiple nodes if the scheduler proves to be problematic.
+    pub async fn run(&self) -> Result<()> {
+        // generate brupop cron expression schedule
+        let brupop_scheduler = BrupopCronScheduler::new()
+            .context(controllerclient_error::GetCronScheduleSnafu)?
+            .generate_schedule()
+            .await
+            .context(controllerclient_error::GetCronScheduleSnafu)?;
+
+        // TODO: this will be removed when we remove update time window settings
+        let brupop_schedule_default = Schedule::from_str(SCHEDULER_DEFAULT).context(
+            controllerclient_error::GenerateScheduleFailedSnafu {
+                cron: SCHEDULER_DEFAULT,
+            },
+        )?;
+
+        // generate brupop update time window schedule
+        // TODO: this will be removed when we remove update time window settings
+        let update_window_start = read_env_var(UPDATE_WINDOW_START_ENV_VAR)?;
+        let update_window_stop = read_env_var(UPDATE_WINDOW_STOP_ENV_VAR)?;
+
+        // When users still set up time window settings and not use new setting cron expression scheduler, we should
+        // run controller with old setting update time window. Otherwise, we should always prioritize to use cron expression scheduler.
+        // TODO: this will be removed when we remove update time window settings
+        if brupop_scheduler.eq(&brupop_schedule_default) {
+            if update_window_start != UPDATE_WINDOW_START_DEFAULT.to_string()
+                || update_window_stop != UPDATE_WINDOW_STOP_DEFAULT.to_string()
+            {
```

Comment:
Rather than compare to the default values, I think it would be better to check whether these values are set at all, it doesn't seem right to allow both to be set.

Reply:
+1 - we should exit early (and document) that only one setting can be set at a time.